src/python/pants/backend/k8s/goals/deploy.py

```diff
@@ -0,0 +1,165 @@
+class KubectlApply:
+    paths: tuple[str, ...]
+    input_digest: Digest
+    platform: Platform
+    env: FrozenDict[str, str] | None = None
+    context: str | None = None
+
+
+@rule(desc="Run k8s deploy process", level=LogLevel.DEBUG)
+async def run_k8s_deploy(
+    field_set: DeployK8sBundleFieldSet,
+    kubectl: Kubectl,
+    k8s_subsystem: K8sSubsystem,
+    platform: Platform,
+) -> DeployProcess:
+    context = field_set.context.value
+    if context is None:
+        raise ValueError(f"Missing `{K8sBundleContextField.alias}` field")
+
+    context = context if kubectl.pass_context else None
+    if context is not None and context not in k8s_subsystem.available_contexts:
+        raise ValueError(
+            f"Context `{context}` is not listed in `[{K8sSubsystem.options_scope}].available_contexts`"
+        )
+
+    dependencies = await Get(
+        Targets, UnparsedAddressInputs, field_set.sources.to_unparsed_address_inputs()
+    )
+    file_sources = await MultiGet(
+        Get(
+            HydratedSources,
+            HydrateSourcesRequest(
+                t.get(SourcesField),
+                for_sources_types=(K8sSourceField,),
+                enable_codegen=True,
+            ),
+        )
+        for t in dependencies
+    )
+    snapshot, target_dependencies = await MultiGet(
+        Get(
+            Snapshot,
+            MergeDigests((*(sources.snapshot.digest for sources in file_sources),)),
+        ),
+        Get(Targets, DependenciesRequest(field_set.dependencies)),
+    )
+
+    publish_targets = [
+        tgt for tgt in target_dependencies if DockerPackageFieldSet.is_applicable(tgt)
+    ]
+
+    env = await Get(
+        EnvironmentVars,
+        EnvironmentVarsRequest,
+        EnvironmentVarsRequest(requested=kubectl.extra_env_vars),
+    )
+
+    process = InteractiveProcess.from_process(
+        await Get(
+            Process,
+            KubectlApply(
+                snapshot.files,
+                platform=platform,
+                input_digest=snapshot.digest,
+                env=env,
+                context=context,
+            ),
+        )
+    )
+
+    description = f"context {context}" if context is not None else None
+
+    return DeployProcess(
+        name=field_set.address.spec,
+        publish_dependencies=tuple(publish_targets),
+        process=process,
+        description=description,
+    )
+
+
+@rule
+async def kubectl_apply_process(
+    request: KubectlApply, platform: Platform, kubectl: Kubectl
+) -> Process:
+    argv: tuple[str, ...] = (kubectl.generate_exe(platform),)
+
+    if request.context is not None:
+        argv += ("--context", request.context)
+
+    argv += ("apply", "-o", "yaml")
+
+    for path in request.paths:
+        argv += ("-f", path)
+
+    kubectl_tool = await Get(
+        DownloadedExternalTool, ExternalToolRequest, kubectl.get_request(platform)
+    )
+    digest = await Get(Digest, MergeDigests([kubectl_tool.digest, request.input_digest]))
```

Comment:
you could use `Process.immutable_input_digest` for the tool. Here's how it's done in the Helm backend https://github.com/pantsbuild/pants/blob/1d1e93edcdf617c651c3eb1d1cbadd29d99172b2/src/python/pants/backend/helm/util_rules/tool.py#L454-L457

Reply:
Sure https://github.com/pantsbuild/pants/pull/21796/commits/72722dd3815ebbf0869c913eec9170edbafb4b25